File: src/sdk/python/ratel_ai/observability/rollup.py
"""Usage-rollup assembly for the lean cloud client (ADR-0013).

A *rollup* is one agent interaction's token accounting — exactly the body the
cloud's `POST /api/v1/events` accepts and the dashboard renders. The numbers
themselves (token estimation, savings, cost) come from `ratel-ai-core` via the
native binding; this module only normalizes and shapes them for the wire.
"""

from __future__ import annotations

import json
from collections.abc import Mapping
from datetime import datetime
from typing import Any

from .. import _native

# The context sources the dashboard breaks spend/savings down by. Order matters
# only for readability; the cloud keys on the names.
CONTEXT_SOURCES = ("skills", "tools", "history", "memory", "user_input")

SourceMap = Mapping[str, Any]


def normalize_sources(value: SourceMap | None) -> dict[str, int] | None:
    """Coerce a (possibly partial) per-source mapping to all five keys as ints,
    missing keys zero-filled. ``None`` passes through so the caller can omit the
    field entirely (e.g. no savings recorded for this interaction)."""
    if value is None:
        return None
    out = {key: 0 for key in CONTEXT_SOURCES}
    for key in CONTEXT_SOURCES:
        raw = value.get(key, 0)
        # Clamp to non-negative — the cloud rejects negative counts, and this keeps
        # parity with the TS client's `raw > 0 ? trunc : 0`.
        out[key] = int(raw) if raw and raw > 0 else 0
    return out


def count_segment(seg: Any) -> int:
    """Token-count one raw context segment via the core estimator: a string
    directly, a list/tuple element-wise, any other object by its compact JSON
    (mirrors the TS client's `estimateTokens(JSON.stringify(seg))`)."""
    if seg is None:
        return 0
    if isinstance(seg, str):
        return int(_native.estimate_tokens(seg))
    if isinstance(seg, (list, tuple)):
        return sum(count_segment(item) for item in seg)
    return int(_native.estimate_tokens(json.dumps(seg, separators=(",", ":"), default=str)))


def tokens_from_context(context: SourceMap) -> dict[str, int]:
    """Derive per-source token counts from raw context segments — pass what you
    already have (the system/skills text, the tools list, the prior messages, the
    retrieved memory, the user's turn) and let the core estimator count each."""
    return {key: count_segment(context.get(key)) for key in CONTEXT_SOURCES}


def build_rollup(
    *,
    tokens_by_category: SourceMap | None = None,
    context: SourceMap | None = None,
    saved_by_category: SourceMap | None = None,
    saveable_by_category: SourceMap | None = None,
    input_tokens: int | None = None,
    output_tokens: int | None = None,
    model: str | None = None,
    latency_ms: int | None = None,
    cost_usd: float | None = None,
    occurred_at: datetime | str | None = None,
) -> dict[str, Any]:
    """Assemble one interaction's rollup event.

    Provide the per-source spend one of two ways: ``tokens_by_category`` when you
    already have exact counts (provider usage, tiktoken), or ``context`` — raw
    segments the SDK token-counts for you (no manual tokenization). When both are
    given, ``tokens_by_category`` wins. ``input_tokens`` defaults to the sum of
    the per-source spend, and ``cost_usd`` is estimated from the model + tokens
    (via the core) unless given.
    """
    per_source = tokens_by_category
    if per_source is None and context is not None:
        per_source = tokens_from_context(context)
    tokens = normalize_sources(per_source) or {key: 0 for key in CONTEXT_SOURCES}
    total = sum(tokens.values())
    event: dict[str, Any] = {"tokens_by_category": tokens}

    saved = normalize_sources(saved_by_category)
    if saved is not None:
        event["saved_by_category"] = saved
    saveable = normalize_sources(saveable_by_category)
    if saveable is not None:
        event["saveable_by_category"] = saveable

    event["input_tokens"] = int(input_tokens) if input_tokens is not None else total
    if output_tokens is not None:
        event["output_tokens"] = int(output_tokens)
    if model:
        event["model"] = model
    if latency_ms is not None:
        event["latency_ms"] = int(latency_ms)

    if cost_usd is not None:
        event["cost_usd"] = float(cost_usd)
    elif model:
        # Cost maths live in the core — estimate from model + tokens.
        out_tokens = int(event.get("output_tokens", 0))
        event["cost_usd"] = round(
            _native.estimate_cost_usd(model, event["input_tokens"], out_tokens), 6
        )

    if occurred_at is not None:
        event["occurred_at"] = (
            occurred_at.isoformat() if isinstance(occurred_at, datetime) else str(occurred_at)
        )
    return event

File: src/sdk/ts/src/rollup.ts
import { estimateCostUsd, estimateTokens } from "../native/index.cjs";

/**
 * Usage-rollup assembly and the metrics-export seam (ADR-0013). Turns one agent
 * interaction into the snake_case *rollup* that Ratel's dashboard ingests at
 * `POST {host}/api/v1/events`. The token / cost maths come from `ratel-ai-core`
 * (native). This module only builds rollups and defines the `Transport` seam a
 * shipper plugs into — the concrete cloud client lives in `@ratel-ai/cloud`.
 */

/** The context sources the dashboard breaks spend and savings down by. */
export const CONTEXT_SOURCES = ["skills", "tools", "history", "memory", "user_input"] as const;

export type ContextSource = (typeof CONTEXT_SOURCES)[number];

/** A full per-source token map (all five keys present). */
export type SourceTokens = Record<ContextSource, number>;

/** A partial per-source map — callers usually set only the sources that apply. */
export type PartialSources = Partial<Record<ContextSource, number>>;

/** A raw context segment to be token-counted for you: a string, an object (a tools
 * array, a message, etc.), or an array of either. */
export type RawSegment = string | object;

/** Raw per-source context. Pass what you already have — the system/skills text, the
 * tools array, the prior messages, the retrieved memory, the user's turn — and the
 * SDK counts the tokens for you via the core estimator. No manual tokenization. */
export interface InteractionContext {
  skills?: RawSegment;
  tools?: RawSegment;
  history?: RawSegment;
  memory?: RawSegment;
  userInput?: RawSegment;
}

/** One interaction's usage, in idiomatic camelCase. */
export interface TrackInput {
  /** Pre-counted per-source spend. Use when you already have exact token counts
   * (provider usage, tiktoken). Otherwise pass `context` and let the SDK count. */
  tokensByCategory?: PartialSources;
  /** Raw context segments — the SDK token-counts each for you. Ignored when
   * `tokensByCategory` is given. */
  context?: InteractionContext;
  /** What Ratel selection kept OUT of the prompt this run. */
  savedByCategory?: PartialSources;
  /** What selection COULD save in observe-only mode. */
  saveableByCategory?: PartialSources;
  inputTokens?: number;
  outputTokens?: number;
  model?: string;
  latencyMs?: number;
  /** Explicit cost; estimated from `model` + tokens when omitted. */
  costUsd?: number;
  occurredAt?: Date | string;
}

/** The on-the-wire rollup — snake_case keys exactly as the cloud accepts them. */
export type Rollup = Record<string, unknown>;

function zeroSources(): SourceTokens {
  return { skills: 0, tools: 0, history: 0, memory: 0, user_input: 0 };
}

function normalizeSources(value: PartialSources | undefined): SourceTokens | undefined {
  if (value == null) return undefined;
  const out = zeroSources();
  for (const key of CONTEXT_SOURCES) {
    const raw = value[key];
    out[key] = raw && raw > 0 ? Math.trunc(raw) : 0;
  }
  return out;
}

function total(sources: SourceTokens): number {
  let sum = 0;
  for (const key of CONTEXT_SOURCES) sum += sources[key];
  return sum;
}

/** Token-count a raw segment via the core estimator: a string directly, an array
 * element-wise, any other object by its compact JSON. */
function countSegment(seg: RawSegment | undefined): number {
  if (seg == null) return 0;
  if (typeof seg === "string") return estimateTokens(seg);
  if (Array.isArray(seg)) return seg.reduce((sum: number, item) => sum + countSegment(item), 0);
  return estimateTokens(JSON.stringify(seg));
}

/** Derive per-source token counts from raw context segments. */
function tokensFromContext(ctx: InteractionContext): PartialSources {
  return {
    skills: countSegment(ctx.skills),
    tools: countSegment(ctx.tools),
    history: countSegment(ctx.history),
    memory: countSegment(ctx.memory),
    user_input: countSegment(ctx.userInput),
  };
}

/**
 * Assemble one interaction's rollup event. The only required field is
 * `tokensByCategory`; `inputTokens` defaults to the per-source sum, and `costUsd`
 * is estimated in-core from model + tokens unless supplied.
 */
export function buildRollup(input: TrackInput): Rollup {
  const perSource =
    input.tokensByCategory ?? (input.context ? tokensFromContext(input.context) : undefined);
  const tokens = normalizeSources(perSource) ?? zeroSources();
  const event: Rollup = { tokens_by_category: tokens };

  const saved = normalizeSources(input.savedByCategory);
  if (saved) event.saved_by_category = saved;
  const saveable = normalizeSources(input.saveableByCategory);
  if (saveable) event.saveable_by_category = saveable;

  const inputTokens = input.inputTokens ?? total(tokens);
  event.input_tokens = inputTokens;
  if (input.outputTokens != null) event.output_tokens = input.outputTokens;
  if (input.model) event.model = input.model;
  if (input.latencyMs != null) event.latency_ms = input.latencyMs;

  if (input.costUsd != null) {
    event.cost_usd = input.costUsd;
  } else if (input.model) {
    const out = input.outputTokens ?? 0;
    event.cost_usd = Math.round(estimateCostUsd(input.model, inputTokens, out) * 1e6) / 1e6;
  }

  if (input.occurredAt != null) {
    event.occurred_at =
      input.occurredAt instanceof Date ? input.occurredAt.toISOString() : String(input.occurredAt);
  }
  return event;
}

/** Sends a batch of rollups somewhere; implemented by `@ratel-ai/cloud` or a custom transport. */
export type Transport = (batch: ReadonlyArray<Rollup>) => Promise<void> | void;
